utils/js/LogUtils.ts

```diff
@@ -103,7 +121,51 @@ function parseSource(source) {
     return '';
 }
 
-function writeLog(msg, source, level) {
-    if (source) msg = `[${source}] ${msg}`;
-    level === 'info' ? console.log(msg) : console.debug(msg);
+function writeLog(msgs: Some<any>, source, level: LogLevel) {
+    msgs = castArray(msgs);
+    if (source) msgs = [`[${source}]`, ...msgs];
+
+    const logArgs = [];
+    let strMsg;
+
+    msgs.forEach((msg, idx) => {
```

Comment:
I think this all would be a bit easier to build logArgs with reduce.  With every step you get the current state of the array and then the next element. For each step,  either push a "|" and your arg, or just append your arg to the string that is the last element

Reply:
Sorry couldn't follow, tried but did not find an easier path with reduce. I tend not to reach for reduce as I almost never find it intuitive or easy to read. More than happy to have a different impl here, just ran out of time to figure it out myself.